File: sdk/rust/src/dstack_client.rs
use anyhow::Result;
use hex::{encode as hex_encode, FromHexError};
use http_client_unix_domain_socket::{ClientUnix, Method};
use reqwest::Client;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use serde_json::{from_str, json, Value};
use sha2::Digest;
use std::collections::HashMap;
use std::env;

const INIT_MR: &str = "000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000";

fn replay_rtmr(history: Vec<String>) -> Result<String, FromHexError> {
    if history.is_empty() {
        return Ok(INIT_MR.to_string());
    }
    let mut mr = hex::decode(INIT_MR)?;
    for content in history {
        let mut content_bytes = hex::decode(content)?;
        if content_bytes.len() < 48 {
            content_bytes.resize(48, 0);
        }
        mr.extend_from_slice(&content_bytes);
        mr = sha2::Sha384::digest(&mr).to_vec();
    }
    Ok(hex_encode(mr))
}

fn get_endpoint(endpoint: Option<&str>) -> String {
    if let Some(e) = endpoint {
        return e.to_string();
    }
    if let Ok(sim_endpoint) = env::var("DSTACK_SIMULATOR_ENDPOINT") {
        return sim_endpoint;
    }
    "/var/run/dstack.sock".to_string()
}

#[derive(Debug)]
pub enum ClientKind {
    Http,
    Unix,
}

#[derive(Serialize, Deserialize)]
pub struct EventLog {
    imr: u32,
    event_type: u32,
    digest: String,
    event: String,
    event_payload: String,
}

#[derive(bon::Builder, Serialize)]
pub struct TlsKeyConfig {
    #[builder(into, default = String::new())]
    pub subject: String,
    #[builder(default = Vec::new())]
    pub alt_names: Vec<String>,
    #[builder(default = false)]
    pub usage_ra_tls: bool,
    #[builder(default = true)]
    pub usage_server_auth: bool,
    #[builder(default = false)]
    pub usage_client_auth: bool,
}

#[derive(Serialize, Deserialize)]
pub struct GetKeyResponse {
    pub key: String,
    pub signature_chain: Vec<String>,
}

impl GetKeyResponse {
    pub fn decode_key(&self) -> Result<Vec<u8>, FromHexError> {
        hex::decode(&self.key)
    }

    pub fn decode_signature_chain(&self) -> Result<Vec<Vec<u8>>, FromHexError> {
        self.signature_chain.iter().map(hex::decode).collect()
    }
}

#[derive(Serialize, Deserialize)]
pub struct GetQuoteResponse {
    pub quote: String,
    pub event_log: String,
}

impl GetQuoteResponse {
    pub fn decode_quote(&self) -> Result<Vec<u8>, FromHexError> {
        hex::decode(&self.quote)
    }

    pub fn decode_event_log(&self) -> Result<Vec<EventLog>, serde_json::Error> {
        serde_json::from_str(&self.event_log)
    }

    pub fn replay_rtmrs(&self) -> Result<HashMap<u8, String>> {
        let parsed_event_log: Vec<EventLog> = self.decode_event_log()?;
        let mut rtmrs = HashMap::new();
        for idx in 0..4 {
            let mut history = vec![];
            for event in &parsed_event_log {
                if event.imr == idx {
                    history.push(event.digest.clone());
                }
            }
            rtmrs.insert(idx as u8, replay_rtmr(history)?);
        }
        Ok(rtmrs)
    }
}

#[derive(Serialize, Deserialize)]
pub struct InfoResponse {
    app_id: String,
    instance_id: String,
    app_cert: String,
    tcb_info: TcbInfo,
    app_name: String,
    public_logs: bool,
    public_sysinfo: bool,
    device_id: String,
    mr_aggregated: String,
    os_image_hash: String,
    mr_key_provider: String,
    key_provider_info: String,
    compose_hash: String,
}

impl InfoResponse {
    pub fn validated_from_value(mut obj: Value) -> Result<Self, serde_json::Error> {
        if let Some(tcb_info_str) = obj.get("tcb_info").and_then(Value::as_str) {
            let parsed_tcb_info: TcbInfo = from_str(tcb_info_str)?;
            obj["tcb_info"] = serde_json::to_value(parsed_tcb_info)?;
        }
        serde_json::from_value(obj)
    }
}

#[derive(Serialize, Deserialize)]
struct TcbInfo {
    mrtd: String,
    rootfs_hash: String,
    rtmr0: String,
    rtmr1: String,
    rtmr2: String,
    rtmr3: String,
    event_log: Vec<EventLog>,
}

#[derive(Serialize, Deserialize)]
pub struct GetTlsKeyResponse {
    pub key: String,
    pub certificate_chain: Vec<String>,
}

pub trait BaseClient {}

pub struct DstackClient {
    base_url: String,
    endpoint: String,
    client: ClientKind,
}

impl BaseClient for DstackClient {}

impl DstackClient {
    pub fn new(endpoint: Option<&str>) -> Self {
        let endpoint = get_endpoint(endpoint);
        let (base_url, client) = match endpoint {
            ref e if e.starts_with("http://") || e.starts_with("https://") => {
                (e.to_string(), ClientKind::Http)
            }
            _ => ("http://localhost".to_string(), ClientKind::Unix),
        };

        DstackClient {
            base_url,
            endpoint,
            client,
        }
    }

    async fn send_rpc_request<S: Serialize, D: DeserializeOwned>(
        &self,
        path: &str,
        payload: &S,
    ) -> anyhow::Result<D> {
        match &self.client {
            ClientKind::Http => {
                let client = Client::new();
                let url = format!(
                    "{}/{}",
                    self.base_url.trim_end_matches('/'),
                    path.trim_start_matches('/')
                );
                let res = client
                    .post(&url)
                    .json(payload)
                    .header("Content-Type", "application/json")
                    .send()
                    .await?
                    .error_for_status()?;
                Ok(res.json().await?)
            }
            ClientKind::Unix => {
                let mut unix_client = ClientUnix::try_new(&self.endpoint).await?;
                let res = unix_client
                    .send_request_json::<_, _, Value>(
                        path,
                        Method::POST,
                        &[("Content-Type", "application/json")],
                        Some(&payload),
                    )
                    .await?;
                Ok(res.1)
            }
        }
    }

    pub async fn get_key(
        &self,
        path: Option<String>,
        purpose: Option<String>,
    ) -> Result<GetKeyResponse> {
        let data = json!({
            "path": path.unwrap_or_default(),
            "purpose": purpose.unwrap_or_default(),
        });
        let response = self.send_rpc_request("/GetKey", &data).await?;
        let response = serde_json::from_value::<GetKeyResponse>(response)?;

        Ok(response)
    }

    pub async fn get_quote(&self, report_data: Vec<u8>) -> Result<GetQuoteResponse> {
        if report_data.is_empty() || report_data.len() > 64 {
            anyhow::bail!("Invalid report data length")
        }
        let hex_data = hex_encode(report_data);
        let data = json!({ "report_data": hex_data });
        let response = self.send_rpc_request("/GetQuote", &data).await?;
        let response = serde_json::from_value::<GetQuoteResponse>(response)?;

        Ok(response)
    }

    pub async fn info(&self) -> Result<InfoResponse> {
        let response = self.send_rpc_request("/Info", &json!({})).await?;
        Ok(InfoResponse::validated_from_value(response)?)
    }

    pub async fn emit_event(&self, event: String, payload: Vec<u8>) -> Result<()> {
        if event.is_empty() {
            anyhow::bail!("Event name cannot be empty")
        }
        let hex_payload = hex_encode(payload);
        let data = json!({ "event": event, "payload": hex_payload });
        self.send_rpc_request::<_, ()>("/EmitEvent", &data).await?;
        Ok(())
    }

    pub async fn get_tls_key(&self, tls_key_config: TlsKeyConfig) -> Result<GetTlsKeyResponse> {
        let response = self.send_rpc_request("/GetTlsKey", &tls_key_config).await?;
        let response = serde_json::from_value::<GetTlsKeyResponse>(response)?;

        Ok(response)
    }
}

File: sdk/python/src/dstack_sdk/dstack_client.py
from typing import Literal, Optional, List, Dict, Any
import binascii
import json
import hashlib
import os
import logging
import base64

from pydantic import BaseModel
import httpx

logger = logging.getLogger('dstack_sdk')


INIT_MR = "000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"


def replay_rtmr(history: list[str]):
    """
    Replay the RTMR history to calculate the final RTMR value.
    """
    if len(history) == 0:
        return INIT_MR
    mr = bytes.fromhex(INIT_MR)
    for content in history:
        # mr = sha384(concat(mr, content))
        # if content is shorter than 48 bytes, pad it with zeros
        content = bytes.fromhex(content)
        if len(content) < 48:
            content = content.ljust(48, b'\0')
        mr = hashlib.sha384(mr + content).digest()
    return mr.hex()


def get_endpoint(endpoint: str | None = None) -> str:
    if endpoint:
        return endpoint
    if "DSTACK_SIMULATOR_ENDPOINT" in os.environ:
        logger.info(
            f"Using simulator endpoint: {os.environ['DSTACK_SIMULATOR_ENDPOINT']}")
        return os.environ["DSTACK_SIMULATOR_ENDPOINT"]
    return "/var/run/dstack.sock"


class GetTlsKeyResponse(BaseModel):
    key: str
    certificate_chain: List[str]


class GetKeyResponse(BaseModel):
    key: str
    signature_chain: List[str]

    def decode_key(self) -> bytes:
        return bytes.fromhex(self.key)

    def decode_signature_chain(self) -> List[bytes]:
        return [bytes.fromhex(chain) for chain in self.signature_chain]


class GetQuoteResponse(BaseModel):
    quote: str
    event_log: str

    def decode_quote(self) -> bytes:
        return bytes.fromhex(self.quote)

    def decode_event_log(self) -> 'List[EventLog]':
        return [EventLog(**event) for event in json.loads(self.event_log)]

    def replay_rtmrs(self) -> Dict[int, str]:
        # NOTE: before dstack-0.3.0, event log might not a JSON file.
        parsed_event_log = json.loads(self.event_log)
        rtmrs = {}
        for idx in range(4):
            history = []
            for event in parsed_event_log:
                if event.get('imr') == idx:
                    history.append(event['digest'])
            rtmrs[idx] = replay_rtmr(history)
        return rtmrs


class EventLog(BaseModel):
    imr: int
    event_type: int
    digest: str
    event: str
    event_payload: str


class TcbInfo(BaseModel):
    mrtd: str
    rootfs_hash: str
    rtmr0: str
    rtmr1: str
    rtmr2: str
    rtmr3: str
    event_log: List[EventLog]


class InfoResponse(BaseModel):
    app_id: str
    instance_id: str
    app_cert: str
    tcb_info: TcbInfo
    app_name: str
    public_logs: bool
    public_sysinfo: bool
    device_id: str
    mr_aggregated: str
    os_image_hash: str
    mr_key_provider: str
    key_provider_info: str
    compose_hash: str

    @classmethod
    def model_validate(cls, obj: Any) -> 'InfoResponse':
        if isinstance(obj, dict) and 'tcb_info' in obj and isinstance(obj['tcb_info'], str):
            obj = dict(obj)
            obj['tcb_info'] = TcbInfo(**json.loads(obj['tcb_info']))
        return super().model_validate(obj)


class BaseClient:
    pass


class DstackClient(BaseClient):
    def __init__(self, endpoint: str | None = None):
        endpoint = get_endpoint(endpoint)
        if endpoint.startswith("http://") or endpoint.startswith('https://'):
            self.transport = httpx.HTTPTransport()
            self.base_url = endpoint
        else:
            self.transport = httpx.HTTPTransport(uds=endpoint)
            self.base_url = "http://localhost"

    def _send_rpc_request(self, path, payload):
        with httpx.Client(transport=self.transport, base_url=self.base_url) as client:
            response = client.post(
                path,
                json=payload,
                headers={"Content-Type": "application/json"}
            )
            response.raise_for_status()
            return response.json()

    def get_key(
        self,
        path: str | None = None,
        purpose: str | None = None,
    ) -> GetKeyResponse:
        """
        Derives a key from the given path + app root key.

        Args:
            path: Path for key derivation (optional)
            purpose: Purpose for key derivation (optional)

        Returns:
            GetKeyResponse object containing key and signature chain

        The key derivation process works as follows:

        1. The app's root key is used as the base for derivation
        2. The provided path is used as context data for HKDF-SHA256 derivation
        3. A derived ECDSA key is created
        4. A message is formed by combining the purpose and the hex-encoded public key of the derived key
        5. The app's key signs this message using Keccak256 (Ethereum-style signing)
        6. The signature includes the recovery ID for verification
        7. A signature chain is returned containing:
           - First element: The app key's signature over the derived key's public key
           - Second element: The KMS root key's signature over the app's public key

        To verify the key:
        1. Verify the app's signature in the chain using the app's public key
        2. Verify the KMS root key's signature over the app's public key
        3. Verify that the message signed by the app key matches the derived key's public key
        4. Optionally, recover the app's public key from its signature and verify it matches
           the expected app public key
        """
        data: Dict[str, Any] = {"path": path or '', "purpose": purpose or ''}
        result = self._send_rpc_request("/GetKey", data)
        return GetKeyResponse(**result)

    def get_quote(
        self,
        report_data: str | bytes,
    ) -> GetQuoteResponse:
        if not report_data or not isinstance(report_data, (bytes, str)):
            raise ValueError("report_data can not be empty")
        is_str = isinstance(report_data, str)
        if is_str:
            report_data = report_data.encode()
        if len(report_data) > 64:
            raise ValueError("report_data must be less than 64 bytes")
        hex = binascii.hexlify(report_data).decode()
        result = self._send_rpc_request("/GetQuote", {"report_data": hex})
        return GetQuoteResponse(**result)

    def info(self) -> InfoResponse:
        result = self._send_rpc_request("/Info", {})
        return InfoResponse.model_validate(result)

    def emit_event(
        self,
        event: str,
        payload: str | bytes,
    ) -> None:
        """
        Emit an event. This extends the event to RTMR3 on TDX platform.

        Requires Dstack OS 0.5.0 or later.

        Args:
            event: The event name
            payload: The event data as string or bytes

        Returns:
            None
        """
        if not event:
            raise ValueError("event name cannot be empty")

        if isinstance(payload, str):
            payload = payload.encode()

        hex_payload = binascii.hexlify(payload).decode()
        self._send_rpc_request("/EmitEvent", {"event": event, "payload": hex_payload})
        return None

    def get_tls_key(
        self,
        subject: str | None = None,
        alt_names: List[str] | None = None,
        usage_ra_tls: bool = False,
        usage_server_auth: bool = False,
        usage_client_auth: bool = False,
    ) -> GetTlsKeyResponse:
        """
        Gets a TLS key from the Dstack service with optional parameters.

        Args:
            subject: The subject for the TLS key
            alt_names: Alternative names for the TLS key
            usage_ra_tls: Whether to enable RA TLS usage
            usage_server_auth: Whether to enable server auth usage
            usage_client_auth: Whether to enable client auth usage

        Returns:
            GetTlsKeyResponse object containing the key and certificate chain
        """
        data: Dict[str, Any] = {
            "subject": subject or "",
            "usage_ra_tls": usage_ra_tls,
            "usage_server_auth": usage_server_auth,
            "usage_client_auth": usage_client_auth,
        }
        if alt_names:
            data["alt_names"] = alt_names

        result = self._send_rpc_request("/GetTlsKey", data)
        return GetTlsKeyResponse(**result)


class AsyncDstackClient(BaseClient):
    def __init__(self, endpoint=None):
        endpoint = get_endpoint(endpoint)
        if endpoint.startswith("http://") or endpoint.startswith('https://'):
            self.transport = httpx.AsyncHTTPTransport()
            self.base_url = endpoint
        else:
            self.transport = httpx.AsyncHTTPTransport(uds=endpoint)
            self.base_url = "http://localhost"

    async def _send_rpc_request(self, path, payload):
        async with httpx.AsyncClient(transport=self.transport, base_url=self.base_url) as client:
            response = await client.post(
                path,
                json=payload,
                headers={"Content-Type": "application/json"}
            )
            response.raise_for_status()
            return response.json()

    async def get_key(
        self,
        path: str | None = None,
        purpose: str | None = None,
    ) -> GetKeyResponse:
        data: Dict[str, Any] = {"path": path or '', "purpose": purpose or ''}
        result = await self._send_rpc_request("/GetKey", data)
        return GetKeyResponse(**result)

    async def get_quote(
        self,
        report_data: str | bytes,
    ) -> GetQuoteResponse:
        if not report_data or not isinstance(report_data, (bytes, str)):
            raise ValueError("report_data can not be empty")
        is_str = isinstance(report_data, str)
        if is_str:
            report_data = report_data.encode()
        if len(report_data) > 64:
            raise ValueError("report_data must be less than 64 bytes")
        hex = binascii.hexlify(report_data).decode()
        result = await self._send_rpc_request("/GetQuote", {"report_data": hex})
        return GetQuoteResponse(**result)

    async def info(self) -> InfoResponse:
        result = await self._send_rpc_request("/Info", {})
        return InfoResponse.model_validate(result)

    async def emit_event(
        self,
        event: str,
        payload: str | bytes,
    ) -> None:
        """
        Emit an event. This extends the event to RTMR3 on TDX platform.

        Requires Dstack OS 0.5.0 or later.

        Args:
            event: The event name
            payload: The event data as string or bytes

        Returns:
            None
        """
        if not event:
            raise ValueError("event name cannot be empty")

        if isinstance(payload, str):
            payload = payload.encode()

        hex_payload = binascii.hexlify(payload).decode()
        await self._send_rpc_request("/EmitEvent", {"event": event, "payload": hex_payload})
        return None

    async def get_tls_key(
        self,
        subject: str | None = None,
        alt_names: List[str] | None = None,
        usage_ra_tls: bool = False,
        usage_server_auth: bool = False,
        usage_client_auth: bool = False,
    ) -> GetTlsKeyResponse:
        """
        Gets a TLS key from the Dstack service with optional parameters.

        Args:
            subject: The subject for the TLS key
            alt_names: Alternative names for the TLS key
            usage_ra_tls: Whether to enable RA TLS usage
            usage_server_auth: Whether to enable server auth usage
            usage_client_auth: Whether to enable client auth usage

        Returns:
            GetTlsKeyResponse object containing the key and certificate chain
        """
        data: Dict[str, Any] = {
            "subject": subject or "",
            "usage_ra_tls": usage_ra_tls,
            "usage_server_auth": usage_server_auth,
            "usage_client_auth": usage_client_auth,
        }
        if alt_names:
            data["alt_names"] = alt_names

        result = await self._send_rpc_request("/GetTlsKey", data)
        return GetTlsKeyResponse(**result)
